Add warning to log, when fpm socket was not registered on the expected path

JoshuaBehrens opened the pull request:
I decided it to be a warning, because you can still run against the newly created socket file but fail it for the configuration test. It felt right to do it like that.

## Where am I coming from?

I have an experiments folder deep down in a folder structure on my work machine (`/home/name/projects/delete-to-free-storage/experiments`). In there I cloned a project where I wanted to try out devenv for a part in a subfolder (`/home/name/projects/delete-to-free-storage/experiments/foobar/php-web`). I set up devenv, it uses process-compose to orchestrate php-fpm for me and it places the php-fpm socket file in a project specific sub directory (`/home/name/projects/delete-to-free-storage/experiments/foobar/php-web/.devenv/state/php-fpm/web`). When I opened up the nginx served page, all I get is a white page and nginx prints into the log `dialing backend: dial unix /home/name/projects/delete-to-free-storage/experiments/foobar/php-web/.devenv/state/php-fpm/web.sock: connect: invalid argument`.

So now try to debug this. You have multiple sources for this error (devenv, process-compose, nix, a nix package for php, php itself, nginx, a nix package for nginx). I was looking into to this with @jochenmanz for several days. Our question was: "Why is the file extension missing?" We were looking through the different tools but no solution

Just today I just received the message "What OS are you on and did you know that socket file names are limited in length?" After just moving the project into a directory with a shorter path, we saw, that it was working and the socket file had the .sock extension as expected. "the extension .sock was missing" is just a coincidence in after a string length cut. So we found the issue it was on binding the socket. We found some checks against in php for every socket access. But for some reason not for php-fpm. The socket will just get cut off. So I decided to fix that so have other people minds less stressed :)

bukka commented on sapi/fpm/fpm/fpm_unix.c:
```diff
@@ -63,6 +64,25 @@ static struct passwd *fpm_unix_get_passwd(struct fpm_worker_pool_s *wp, const ch
+    size_t address_length = strlen(address);
+    size_t socket_length = sizeof(test_socket.sun_path);
+
+    if (wp->listen_address_domain != FPM_AF_UNIX) {
+        return true;
+    }
```
NIT: you can check this first before getting the length - we no longer require definition at the top (C89) as at least c99 is currently required.

JoshuaBehrens replied:
Great to hear :) I was worried I have to support some old compiler. Sure will rewrite that

bukka commented on sapi/fpm/fpm/fpm_unix.c:
```diff
@@ -63,6 +64,25 @@ static struct passwd *fpm_unix_get_passwd(struct fpm_worker_pool_s *wp, const ch
+
+    zlog(flags, "[pool %s] cannot bind to UNIX socket '%s' as path is too long (found length: %lu, maximal length: %lu)", wp->config->name, address, address_length, socket_length);
+
+    return false;
```
Just a note that this sort of makes it slightly inconsistent as the test fails but the actual run might still work trimmed. I guess it probably makes sense to fail the test as most of the time this is not what users would want.

There should be a special test for that.

JoshuaBehrens replied:
Yes, I saw that inconsistency as well. I tried to mention, that in my PR description "I decided it to be a warning, because you can still run against the newly created socket file but fail it for the configuration test. It felt right to do it like that."

bukka replied:
Ok will leave this open as we might need a note in UPGRADING for this which I will add before merging.

bukka commented on sapi/fpm/fpm/fpm_sockets.c:
```diff
@@ -396,9 +396,20 @@ static int fpm_socket_af_inet_listening_socket(struct fpm_worker_pool_s *wp) /*
+    if (address_length >= socket_length) {
+        zlog(
+            ZLOG_WARNING,
+            "[pool %s] cannot bind to UNIX socket '%s' as path is too long (found length: %lu, maximal length: %lu), trying cut socket path instead '%s'",
```
It would be good to try to add a test for this. It should be a bit generic as I think it might differ on FreeBSD so not rely on exact size but use some RegExp.

JoshuaBehrens replied:
`\[pool .+\] cannot bind to UNIX socket '.+' as path is too long \(found length: \d+, maximal length: \d+\), trying cut socket path instead '.+'`
some pattern like this?

bukka commented on sapi/fpm/fpm/fpm_unix.c:
```diff
@@ -63,6 +64,25 @@ static struct passwd *fpm_unix_get_passwd(struct fpm_worker_pool_s *wp, const ch
+        return true;
+    }
+
+    zlog(flags, "[pool %s] cannot bind to UNIX socket '%s' as path is too long (found length: %lu, maximal length: %lu)", wp->config->name, address, address_length, socket_length);
```
NIT: please wrap the params to the new line...

bukka reviewed (changes requested):
It makes sense but needs some tests.

JoshuaBehrens commented:
@bukka I am happy to write some tests for it. Can you give me a reference test that I can copy or extend for my feature?

bukka commented:
Check this one out: https://github.com/php/php-src/blob/414f71a90254cc33896bb3ba953f979f743c198c/sapi/fpm/tests/socket-uds-numeric-ugid-nonroot.phpt . It should hopefully give you some idea. You will need to create a custom address so something with a long suffix should hopefully do.

JoshuaBehrens commented:
Hi @bukka I created two tests. One for actually running fpm and one for the test. I added a section in the upgrade notes. I am up to any wording changes :) With this I noticed I affected other tests as well. For example I broke bug68591-conf-test-listen-group.log:
```
ERROR: [pool unconfined] cannot get gid for group 'aaaaaa': %s
ERROR: FPM initialization failed
Done
---- ACTUAL OUTPUT
ERROR: [pool unconfined] cannot bind to UNIX socket '/var/folders/7j/lfxr9fd12tn1lcfj3bl9f_7r0000gn/T/nix-shell.XXXXXX.mu6K44Ew/5cf922da-bug68591-conf-test-listen-group.9008.sock' as path is too long (found length: 125, maximal length: 104)
ERROR: FPM initialization failed
```
I was able to unbreak it but therefore I changed code in the Tester class. This might affect more than I can grasp right now but it fights the fact on Mac the temp folder is not /tmp and therefore not a shorter but rather longer option for a socket file path.

Sidenote: Are you interested in feedback on my experience as first time contributor?

iluuu1994 commented:
There's https://www.phpinternalsbook.com/, which is likely the most up to date resource atm. But that doesn't help much when it's linked nowhere 😄 Adding a basic contributors guide is certainly a good idea.

JoshuaBehrens commented:
Without having a detailed look into it it looks promising :D I will give it a go next time I am up to change something

iluuu1994 commented:
@JoshuaBehrens I created a PR here: https://github.com/php/php-src/pull/11155 Feel free to comment if you have any other suggestions!

bukka reviewed (changes requested):
The C part looks good. Just the tests need some changes (mostly smaller one - it looks good in general).

bukka commented on UPGRADING:
```diff
@@ -467,6 +467,12 @@ PHP 8.2 UPGRADE NOTES
   Previously only hexadecimal prefixes and using a leading 0 for octal numbers
   was supported.
 
+- Parsing of invalid values will now trigger a warning when this was silently
+  ignored before. Interpretation of these values is not changed, for
+  backwards compatibility. Testing configuration can fail though.
+  This affects the following settings:
+  . {fpm_pool}.listen
+
```
I see some conflicts as this is getting updated by other changes. Feel free to remove it and just leave the text in the comment. It can be added later during merge in the same way like we update NEWS...

JoshuaBehrens commented on sapi/fpm/tests/gh11066-fpm-unix-socket-too-long-filename-but-starts-anyway.phpt:
```diff
@@ -0,0 +1,81 @@
+$tester->expectLogTerminatingNotices();
+$tester->close();
+
+echo ini_get('fpm_pool.listen');
```
Likely debugging :D ty for pointing out

JoshuaBehrens commented on sapi/fpm/tests/gh11066-fpm-unix-socket-too-long-filename-but-starts-anyway.phpt:
```diff
@@ -0,0 +1,81 @@
+// cleanup socket file if php-fpm was not killed
+$tempDir = sys_get_temp_dir();
+$socketFilePrefix = $tempDir . '/socket-file';
+$socketFile = $socketFilePrefix . '-that-is-too-long-0000-1111-2222-3333-4444-5555-6666-7777-8888-9999-aaaa-bbbb-cccc-dddd-eeee-ffff-gggg-hhhh-iiii.sock';
+
+if (is_file($socketFile)) {
```
`is this actually correct if it creates trimmed file?` this is what happens already today. I just try to make one aware of it so one can bind to the right socket

JoshuaBehrens commented on sapi/fpm/tests/tester.inc:
```diff
@@ -1144,16 +1144,22 @@ class Tester
+                   hash('crc32', dirname($address)) . '-' .
+                   basename($address);
+            $addresses = [$address, $tmpAddress];
+            \usort($addresses, function (string $a, string $b): int {
+                return strlen($a) <=> strlen($b);
+            });
 
             // Socket max path length is 108 on Linux and 104 on BSD,
             // so we use the latter
-            if (strlen($address) <= 104) {
-                return $address;
+            if (strlen($addresses[1]) > 104) {
+                return $addresses[0];
             }
 
-            return sys_get_temp_dir() . '/' .
-                   hash('crc32', dirname($address)) . '-' .
-                   basename($address);
+            return $addresses[1];
```
I mentioned it here https://github.com/php/php-src/pull/11066#issuecomment-1525565283 I broke tests when I was testing on a Mac as the temporary path is super long. I moved the code to a location where paths are smaller than the tmp folder. Some tests still failed as they created socket files, that were just too long in the temp folder but they expected to be valid socket files. So I saw how these socket file paths were created (the code above) and I changed its fallback logic to prefer short paths as this was the leading issue for tests to fail for no "good" reason.

JoshuaBehrens commented:
I rebased and amend the code in previous commits so it is a bit cleaner to read the history.

I suggest an UPGRADE notice like:
```
- Parsing of invalid values will now trigger a warning when this was silently
  ignored before. Interpretation of these values is not changed, for
  backwards compatibility. Testing configuration can fail though.
  This affects the following settings:
  . {fpm_pool}.listen
```

bukka commented on sapi/fpm/fpm/fpm_unix.c:
```diff
@@ -88,8 +112,10 @@ bool fpm_unix_test_config(struct fpm_worker_pool_s *wp)
+		fpm_unix_check_listen_address(wp, config->listen_address, ZLOG_ERROR) &&
 		fpm_unix_check_passwd(wp, config->user, ZLOG_ERROR) &&
 		fpm_unix_check_group(wp, config->group, ZLOG_ERROR) &&
+        fpm_unix_check_listen_address(wp, config->listen_address, ZLOG_SYSERROR) &&
```
why is this being done twice?

JoshuaBehrens replied:
You are right. I was adding it twice, one for ZLOG_ERROR and one for ZLOG_SYSERROR as this has been done twice for user and group as well, but I misread the lines. It was checking four different configuration values. So is this rather a ZLOG_ERROR or a ZLOG_SYSERROR? I tend to go for ZLOG_SYSERROR

bukka replied:
Yeah agreed ZLOG_SYSERROR is more consistent...

bukka replied:
I addressed this.

bukka commented on sapi/fpm/tests/gh11066-fpm-unix-socket-too-long-filename-but-starts-anyway.phpt:
```diff
@@ -0,0 +1,66 @@
+<?php
+
+require_once "tester.inc";
+$tempDir = sys_get_temp_dir();
```
it would be better to do it in the test dir and use test naming as it's the case for other tests - it could maybe use some helper function for tester (that's not necessary if you find it too hard but might be nicer).

JoshuaBehrens replied:
I am not sure where a helper method would simplify this so I solved it as suggested

bukka reviewed (commented):
Apology for slightly late review. Completely forgot about this when doing FPM stuff last month... Just some minor suggestions and one question added.

JoshuaBehrens commented:
I rebased, amend previous commits to the given feedback.

I hope I got the helper for the socket directory into a more welcome solution. I need feedback on the log type so I can amend this as well :)

bukka commented:
So I have been testing this and it was actually failing due to incorrect test (the one for FPM test error check) which I needed to rewrite. Also pipeline build was failing because it requires using `%zu` when formatting `size_t` which I also fixed. I did some few other things as well (CS, test formatting and that extra test check). I have done all those changes myself as the feature freeze is on Tuesday so it can get to 8.3. After that pipeline was green so it is all merged now! Thanks for you work!

JoshuaBehrens commented:
Thank you for the collaborative work and investment to get it in before the feature freeze <3

andypost commented:
Trying to build for Alpinelinux getting the test to fail
```
TEST 16510/16569 [sapi/fpm/tests/socket-uds-too-long-filename-start.phpt]
========DIFF========
+ Socket files were not found.
+
+ Warning: Undefined array key 0 in /builds/alpine/aports/testing/php83/src/php-src-0b0cec5b8a920bafa07dd702950a092a8811596d/sapi/fpm/tests/socket-uds-too-long-filename-start.php on line 41
     Done
========DONE========
[FAIL FPM: UNIX socket filename is too for start [sapi/fpm/tests/socket-uds-too-long-filename-start.phpt] 
```

andypost commented:
Running within `/builds/alpine` fails as well as `/mnt0123456789` - so length to source should be less then 88 chars

andypost commented:
That's because socket maxpath is 108 chars
